File: ruby/get_segments.rb
# frozen_string_literal: true

require 'faraday'
require 'json'

BASE_URL = 'https://api.aweber.com/1.0/'
credentials = JSON.parse(File.read('./credentials.json'))

conn = Faraday.new(
  headers: {
    Authorization: "Bearer #{credentials['access_token']}"
  }
) do |f|
  f.response :raise_error
end

def get_collection(conn, url)
  collection = []
  while url
    res = conn.get(url)
    page = JSON.parse(res.body)
    collection.push(*page["entries"])
    url = page["next_collection_link"]
  end
  collection
end

# Get an account to search on
accounts = get_collection(conn, "#{BASE_URL}accounts")

lists_url = accounts[0]['lists_collection_link']
lists = get_collection(conn, lists_url)
list = lists.first

segments = get_collection(conn, list['segments_collection_link'])
puts "segments #{segments}"

segments.each do |segment|
  puts "#{segment['name']} #{segment['self_link']}"
end

File: node/get-segments.js
const fetch = require("node-fetch");
const credentials = require("./credentials.json");
const BASE_URL = 'https://api.aweber.com/1.0/';
const client_id = credentials['clientId'];
const client_secret = credentials['clientSecret'];
const accessToken = credentials['accessToken'];

const qs = require("querystring");

async function request(url, options) {
    if (options == null) options = {};
    if (options.credentials == null) options.credentials = 'same-origin';
    return fetch(url, options).then(function(response) {
        if (response.status >= 200 && response.status < 300) {
            return Promise.resolve(response);
        } else {
            var error = new Error(response.statusText || response.status);
            error.response = response;
            return Promise.reject(error);
        }
    });
}

async function getCollection(accessToken, url) {
    let res;
    const collection = [];
    console.log({
        accessToken
    })
    while (url) {
        res = await request(url, {
            headers: {
                "Authorization": `Bearer ${accessToken}`
            }
        });
        let page = await res.json();
        console.log("got page", {
            page
        })
        collection.push(...page.entries);
        url = page.next_collection_link;
    }
    return collection;
}
(async() => {

    // Get an account to search on
    const accounts = await getCollection(accessToken, BASE_URL + 'accounts');

    // Get a list to find segments on
    const lists = await getCollection(accessToken, accounts[0]['lists_collection_link']);

    // get the segments associated with a list
    const segments = await getCollection(accessToken, lists[0]['segments_collection_link']);
    console.log("segments", segments)
    for (let segment of segments) {
        console.log(`${segment['name']} ${segment['self_link']}`);
    }
})();
